### PR TITLE
Serialize Zipline Manifest

### PR DESCRIPTION
**Pull request by adrw**
Serialize Zipline Manifest

### CODE REVIEW

File: zipline-gradle-plugin/src/test/kotlin/app/cash/zipline/gradle/ZiplineCompileTaskTest.kt [hunk 1]
```diff
@@ -16,9 +16,9 @@
 
 package app.cash.zipline.gradle
 
-import app.cash.zipline.CURRENT_ZIPLINE_VERSION
 import app.cash.zipline.QuickJs
-import app.cash.zipline.ZiplineFile
+import app.cash.zipline.loader.CURRENT_ZIPLINE_VERSION
+import app.cash.zipline.loader.ZiplineFile
```
Comment (adrw) [inline zipline-gradle-plugin/src/test/kotlin/app/cash/zipline/gradle/ZiplineCompileTaskTest.kt:21]: We moved `ZiplineFile` to `zipline-loader` since it better fits there

File: zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+package app.cash.zipline.loader
+
+import kotlinx.serialization.Serializable
+
+// Constructor is still accesible from .copy() on data class, alternatively can implement without the data class but
+// with manual equals/hashCode to support value equality
+@Serializable
+data class ZiplineManifest private constructor(
```
Comment (adrw) [inline zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt:26]: We split out many of these classes in the loader package into their own files

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+          dependsOnIds = listOf("alpha"),
+        ),
+      )
+    )
+
+    val serialized = Json { prettyPrint = true }.encodeToString(original)
+    assertThat(serialized).isEqualTo(
+      """
+      |{
+      |    "modules": {
```
Comment (adrw) [inline zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:142]: Manifest serialization works 🎉

File: zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.zipline.loader
+
+import kotlinx.serialization.Serializable
+
+// Constructor is still accesible from .copy() on data class, alternatively can implement without the data class but
```
Comment (acostama) [inline zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt]: turn this into java doc? .. eg. point to using ZiplineManifest.create instead of constructor / copy

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+import com.google.common.truth.Truth.assertThat
+import kotlin.test.assertFailsWith
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import okio.ByteString.Companion.encodeUtf8
+import org.junit.Test
+
+// for some reason naming this file ZiplineManifestTest.kt breaks this file in Intellij
```
Comment (acostama) [inline zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:27]: does it break outside of intellij?

**Review by acostama — APPROVED** [review]
overall lgtm

File: zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineLoader.kt [hunk 6]
```diff
@@ -22,20 +37,24 @@ class ZiplineLoader(
...
+      ModuleLoad(zipline, ziplineMutex, concurrentDownloadsSemaphore, it.key, it.value, mutableListOf())
     }
 
-    val idToLoad = loads.associateBy { it.module.id }
+    val idToLoad = loads.associateBy { it.id }
 
     val loadsSorted = loads.topologicalSort { load ->
       load.module.dependsOnIds.map { upstream ->
-        idToLoad[upstream] ?: throw IllegalArgumentException("${load.module.id} depends on unknown module $upstream")
+        idToLoad[upstream] ?: throw IllegalArgumentException("${load.id} depends on unknown module $upstream")
```
Comment (swankjesse) [inline zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineLoader.kt]: skip the topological sort now that the manifest does it?

(might want to move this nice exception to the manifest instead of `!!`)

File: zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+// with manual equals/hashCode to support value equality
+@Serializable
+data class ZiplineManifest private constructor(
+  /** This is an ordered map; its modules are always topologically sorted. */
+  val modules: Map<String, ZiplineModule>
+) {
+  init {
+    require(modules.toList().isTopologicallySorted { (_, module) -> module.sourceToTarget(modules) }) {
```
Comment (swankjesse) [inline zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt]: Great idea

File: zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      "Modules are not topologically sorted and can not be loaded"
+    }
+  }
+
+  companion object {
+    private fun ZiplineModule.sourceToTarget(modules: Map<String, ZiplineModule>) =
+      dependsOnIds.map { dependsOnId -> dependsOnId to modules[dependsOnId]!! }
+
+    fun create(modules: Map<String, ZiplineModule>): ZiplineManifest =
+      ZiplineManifest(modules.toList().topologicalSort { (_, module) -> module.sourceToTarget(modules) }.toMap())
```
Comment (swankjesse) [inline zipline-loader/src/main/kotlin/app/cash/zipline/loader/ZiplineManifest.kt]: This hard to follow and it’s my fault. I have an idea though! What if instead of doing a toplogical sort on the `List<Pair<String, Module>>`, what if we do a topological sort on the IDs, then use those IDs to build a map?

```
val idsSorted = modules.keys.topologicalSort { id ->
  val module = modules[id] ?: 
    throw IllegalArgumentException("unexpected dependency on unknown module $id")
  return@topologicalSort module.dependsOnIds
}
val modulesSorted = idsSorted.associateWith { modules[it] }
return ZiplineManifest(modulesSorted)
```

File: zipline-loader/src/main/kotlin/app/cash/zipline/loader/topologicalSort.kt [hunk 9]
```diff
@@ -48,3 +48,12 @@ fun <T> List<T>.topologicalSort(sourceToTarget: (T) -> Iterable<T>): List<T> {
 
   return result
 }
+
+fun <T> List<T>.isTopologicallySorted(sourceToTarget: (T) -> Iterable<T>): Boolean {
+  val seenNodes = mutableSetOf<T>()
+  for (node in this) {
+    if (!seenNodes.containsAll(sourceToTarget(node).toSet())) return false
```
Comment (swankjesse) [inline zipline-loader/src/main/kotlin/app/cash/zipline/loader/topologicalSort.kt]: save a set allocation:

```
    if (sourceToTarget(node).any { it !in seenNodes }) return false
```

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/TopologicalSortTest.kt [hunk 10]
```diff
@@ -41,8 +57,20 @@ class TopologicalSortTest {
     }
   }
 
+  private fun assertTopologicalSort(unsorted: List<String>, sorted: List<String>, vararg edges: String) {
```
Comment (swankjesse) [inline zipline-loader/src/test/kotlin/app/cash/zipline/loader/TopologicalSortTest.kt:76]: niiiiiice

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/TopologicalSortTest.kt [hunk 11]
```diff
@@ -41,8 +57,20 @@ class TopologicalSortTest {
...
+    val sourceToTarget = edges(*edges)
+    assertThat(unsorted.isTopologicallySorted(sourceToTarget)).isFalse()
+    assertThat(sorted.isTopologicallySorted(sourceToTarget)).isTrue()
+
+    val actual = unsorted
+      .topologicalSort(sourceToTarget)
+
+    assertThat(actual.isTopologicallySorted(sourceToTarget)).isTrue()
```
Comment (swankjesse) [inline zipline-loader/src/test/kotlin/app/cash/zipline/loader/TopologicalSortTest.kt:84]: This is redundant with line 63?

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt [hunk 12]
```diff
@@ -0,0 +1,146 @@
+package app.cash.zipline.loader
```
Comment (swankjesse) [inline zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:17]: nit: copyright

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt (hunk 5 above)
Comment (swankjesse) [reply zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:27]: There’s a git + filesystems thing where case insensitive file systems will fight you if you rename a file by changing its case only.

Comment (swankjesse) [reply zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:27]: Probably works best if you rename to `Foo.kt` and then rename again to the name you want

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+          ),
+          "bravo" to ZiplineModule(
+            url = "/bravo.zipline",
+            sha256 = "abc123".encodeUtf8(),
+            dependsOnIds = listOf("alpha"),
+          ),
+        )
+      )
+    }
+    assertThat(cyclicalException.message).isEqualTo("No topological ordering is possible for [(alpha, ZiplineModule(url=/alpha.zipline, sha256=[text=abc123], patchFrom=null, patchUrl=null, dependsOnIds=[bravo])), (bravo, ZiplineModule(url=/bravo.zipline, sha256=[text=abc123], patchFrom=null, patchUrl=null, dependsOnIds=[alpha]))]")
```
Comment (swankjesse) [inline zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt]: Doing a topological sort on IDs also tightens up this message.

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt (hunk 3 above)
Comment (swankjesse) [reply zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:142]: yay!

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+      |            "dependsOnIds": [
+      |                "alpha"
+      |            ]
+      |        }
+      |    }
+      |}
+    """.trimMargin()
+    )
+
+    val parsed = Json { }.decodeFromString<ZiplineManifest>(serialized)
```
Comment (swankjesse) [inline zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt]: is the `{ }` unnecessary?

**Review by swankjesse — APPROVED** [review]

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt (hunk 5 above)
Comment (JakeWharton) [reply zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:27]: 1. Use a case-sensitive file system for your code!
2. This is not actually a problem in IntelliJ anymore: https://blog.jetbrains.com/idea/2016/06/how-to-support-case-only-rename-in-git/

Comment (adrw) [reply zipline-loader/src/test/kotlin/app/cash/zipline/loader/ziplineManifestTest.kt:27]: I've tried renaming to other things and back, deleting the file, recreating, emptying caches to no avail.

File: zipline-loader/src/test/kotlin/app/cash/zipline/loader/TopologicalSortTest.kt (hunk 11 above)
Comment (adrw) [reply zipline-loader/src/test/kotlin/app/cash/zipline/loader/TopologicalSortTest.kt:84]: L63 asserts that the sorted "expected" list is actually sorted. L68 checks that the sorted output is sorted. Probably redundant but trying to catch any possible unsorted issue
